exporter/exporterhelper/resource_to_label.go

```diff
@@ -0,0 +1,140 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+)
+
+// ResourceToLabelSettings defines configuration for converting resource attributes to labels
+type ResourceToLabelSettings struct {
+	// Enabled indicates whether to not convert resource attributes to labels
+	Enabled bool `mapstructure:"enabled"`
+}
+
+// CreateDefaultResourceToLabelSettings returns the default settings for ResourceToLabelSettings.
+func createDefaultResourceToLabelSettings() ResourceToLabelSettings {
+	return ResourceToLabelSettings{
+		Enabled: false,
+	}
+}
+
+// convertResourceToLabels converts all resource attributes to metric labels
+func convertResourceToLabels(md pdata.Metrics) pdata.Metrics {
+	cloneMd := md.Clone()
+	rms := cloneMd.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		resource := rms.At(i).Resource()
+
+		labelMap := extractLabelsFromResource(&resource)
+
+		ilms := rms.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ilm := ilms.At(j)
+			if ilm.IsNil() {
+				continue
+			}
+			metricSlice := ilm.Metrics()
+			for k := 0; k < metricSlice.Len(); k++ {
+				metric := metricSlice.At(k)
+				if metric.IsNil() {
+					continue
+				}
+				addLabelsToMetric(&metric, labelMap)
+			}
+		}
+	}
+	return cloneMd
+}
+
+// extractAttributesFromResource extracts the attributes from a given resource and
+// returns them as a StringMap.
+func extractLabelsFromResource(resource *pdata.Resource) pdata.StringMap {
+	labelMap := pdata.NewStringMap()
+
+	attrMap := resource.Attributes()
+	attrMap.ForEach(func(k string, av pdata.AttributeValue) {
+		stringLabel := tracetranslator.AttributeValueToString(av, false)
+		labelMap.Insert(k, stringLabel)
```

Comment:
Upsert is a bit faster and here has the same behavior.

Reply:
Updated